File: autonav_ws/autonav_filters/src/filters.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "autonav_filters/filters.hpp"
#include "rclcpp/rclcpp.hpp"
#include "autonav_messages/msg/motor_feedback.hpp"
#include "autonav_messages/msg/gps_feedback.hpp"
#include "autonav_messages/msg/position.hpp"

#define _USE_MATH_DEFINES

using namespace std::chrono_literals;

//particle filter
float latitudeLength = 111086.2;
float longitudeLength = 81978.2;
ParticleFilter particle_filter{latitudeLength, longitudeLength};
autonav_messages::msg::GPSFeedback first_gps;
autonav_messages::msg::GPSFeedback last_gps;
bool first_gps_received = false;

void FiltersNode::on_GPS_received(const autonav_messages::msg::GPSFeedback gps_message) {
    if (gps_message.gps_fix == 0 &&  gps_message.is_locked == false) {
        return;
    }
    if (this->first_gps_received = false) {
        this->first_gps = gps_message;
        this->first_gps_received = true;
    }

    this->last_gps = gps_message;

    this->particle_filter.gps(gps_message);
    
}
void FiltersNode::on_MotorFeedback_received(const autonav_messages::msg::MotorFeedback motorFeedback_message) {
    std::vector<float> averages;
    averages = this->particle_filter.feedback(motorFeedback_message);

    if (averages.size() < 3) {
        return;
    }

    autonav_messages::msg::Position position;
    position.x = averages[0];
    position.y = averages[1];
    position.theta = (-1 * M_PI * 2 + averages[2]);

    if (this->first_gps_received = true) {
        float gps_x = this->first_gps.latitude + position.x / this->latitudeLength;
        float gps_y = this->first_gps.longitude - position.y / this->longitudeLength;

        position.latitude = gps_x;
        position.longitude = gps_y;
    }

    positionPublisher->publish(position);
}

rclcpp::Subscription<autonav_messages::msg::GPSFeedback>::SharedPtr gps_subscription;
rclcpp::Subscription<autonav_messages::msg::MotorFeedback>::SharedPtr motor_subscription;
rclcpp::Publisher<autonav_messages::msg::Position>::SharedPtr positionPublisher;
size_t count_;

File: autonav_ws/src/autonav_filters/src/filters.py
#!/usr/bin/env python3

import json
from types import SimpleNamespace
from autonav_msgs.msg import MotorFeedback, GPSFeedback, Position, IMUData
from scr.states import DeviceStateEnum, SystemStateEnum, SystemModeEnum
from particlefilter import ParticleFilter
from deadrekt import DeadReckoningFilter
from scr_msgs.msg import SystemState
from scr.node import Node
from enum import IntEnum
import rclpy
import math


class FilterType(IntEnum):
    DEAD_RECKONING = 0,
    PARTICLE_FILTER = 1


class FiltersNodeConfig:
    def __init__(self):
        self.filter_type = FilterType.PARTICLE_FILTER


class FiltersNode(Node):
    def __init__(self):
        super().__init__("autonav_filters")

        self.first_gps = None
        self.last_gps = None

        self.latitude_length = self.declare_parameter("latitude_length", 111086.2).get_parameter_value().double_value
        self.longitude_length = self.declare_parameter("longitude_length", 81978.2).get_parameter_value().double_value

        self.pf = ParticleFilter(self.latitude_length, self.longitude_length)
        self.dr = DeadReckoningFilter()
        self.config = self.get_default_config()

        self.onReset()

    def config_updated(self, jsonObject):
        self.config = json.loads(self.jdump(jsonObject), object_hook=lambda d: SimpleNamespace(**d))

    def get_default_config(self):
        return FiltersNodeConfig()

    def init(self):
        self.create_subscription(GPSFeedback, "/autonav/gps", self.onGPSReceived, 20)
        self.create_subscription(MotorFeedback, "/autonav/MotorFeedback", self.onMotorFeedbackReceived, 20)
        self.position_publisher = self.create_publisher(Position, "/autonav/position", 20)

        self.set_device_state(DeviceStateEnum.OPERATING)

    def onReset(self):
        self.dr.reset()
        self.pf.init_particles()

    def system_state_transition(self, old: SystemState, updated: SystemState):
        if old.state != SystemStateEnum.AUTONOMOUS and updated.state == SystemStateEnum.AUTONOMOUS:
            self.onReset()

        if old.mobility == False and updated.mobility == True:
            self.onReset()

    def onGPSReceived(self, msg: GPSFeedback):
        if msg.gps_fix == 0 and msg.is_locked == False:
            return

        if self.first_gps is None:
            self.first_gps = msg

        self.last_gps = msg
        if self.config.filter_type == FilterType.PARTICLE_FILTER:
            self.pf.gps(msg)
        else:
            self.dr.gps(msg)

    def onMotorFeedbackReceived(self, msg: MotorFeedback):
        averages = None
        if self.config.filter_type == FilterType.PARTICLE_FILTER:
            averages = self.pf.feedback(msg)
        else:
            averages = self.dr.feedback(msg)

        if averages is None:
            return

        position = Position()
        position.x = averages[0]
        position.y = averages[1]
        position.theta = (-1 * math.pi * 2 + averages[2]) * 1

        if self.first_gps is not None:
            gps_x = self.first_gps.latitude + position.x / self.latitude_length
            gps_y = self.first_gps.longitude - position.y / self.longitude_length
            position.latitude = gps_x
            position.longitude = gps_y

        self.position_publisher.publish(position)


def main():
    rclpy.init()
    node = FiltersNode()
    Node.run_node(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
